### PR TITLE
8327109: Refactor data graph cloning used in create_new_if_for_predicate() into separate class

### PR DESCRIPTION
**Pull request by chhagedorn**
In the review process of https://github.com/openjdk/jdk/pull/16877, we identified an existing issue in `create_bool_from_template_assertion_predicate()` which is also still present in the refactoring of https://github.com/openjdk/jdk/pull/16877: In rare cases, we could endlessly re-process nodes in the DFS walk since a visited set is missing. This needs to be addressed.

#### Redo refactoring of `create_bool_from_template_assertion_predicate()`
On top of that bug, the refactored version of https://github.com/openjdk/jdk/pull/16877 is still quite complicated to understand since it tries to do multiple steps simultaneously. We've decided to redo the refactoring and better separate the steps to simplify the algorithm. By doing so, we also want to fix the existing bug. This work is split into three separate RFEs (JDK-8327109, JDK-8327110, and JDK-8327111).

#### Share data graph cloning code - start from existing code 
This first PR starts with the existing code found in `clone_nodes_with_same_ctrl()` which is called by `create_new_if_for_predicate()`. `clone_nodes_with_same_ctrl()` already does the data graph cloning in 3 separate steps which can be used as foundation:

1. Collect data nodes to clone by using a node filter
2. Clone the collected nodes (their data and control inputs still point to the old nodes)
3. Fix the cloned data node inputs pointing to the old nodes to the cloned inputs by using an old->new mapping. In this pass, also fix the control inputs of any pinned data node from the old uncommon projection to the new one.

#### Shared data graph cloning class
Some of these steps above are shared with the data graph cloning done in `create_bool_from_template_assertion_predicate()` (refactored in JDK-8327110 and JDK-8327111). We therefore extract them in this patch such that we can reuse it in the refactoring for `create_bool_from_template_assertion_predicate()` later. We create a new `DataNodeGraph` class which does the following (to be shared) cloning of a data graph:

1. Take a collection of data nodes (the collection step is different in `clone_nodes_with_same_ctrl()` compared to `create_bool_from_template_assertion_predicate()` and thus cannot be shared) and clone them. [Same as step 2 above]
 2. Fix the cloned data node inputs pointing to the old nodes to the cloned inputs by using an old->new mapping. [Same as first part of step 3 above but drop the second part of rewiring control inputs which is specific to `clone_nodes_with_same_ctrl()`]

`DataNodeGraph` can then later be reused in JDK-8327110 and JDK-8327111 to refactor `create_bool_from_template_assertion_predicate()`.

Thanks to @eme64 for the comments in https://github.com/openjdk/jdk/pull/16877 and the joint effort to find a reproducer of the existing bug which was the main motivation to redo the refactoring.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327109](https://bugs.openjdk.org/browse/JDK-8327109): Refactor data graph cloning used in create_new_if_for_predicate() into separate class (**Sub-task** - P4)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) ⚠️ Review applies to [14b46ba6](https://git.openjdk.org/jdk/pull/18080/files/14b46ba6f3e2db4acbddb979f0f8d176b476b553)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18080/head:pull/18080` \
`$ git checkout pull/18080`

Update a local copy of the PR: \
`$ git checkout pull/18080` \
`$ git pull https://git.openjdk.org/jdk.git pull/18080/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18080`

View PR using the GUI difftool: \
`$ git pr show -t 18080`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18080.diff">https://git.openjdk.org/jdk/pull/18080.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18080#issuecomment-1973215225)

### CODE REVIEW

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 1]
```diff
@@ -242,61 +244,31 @@ Unique_Node_List PhaseIdealLoop::find_nodes_with_same_ctrl(Node* node, const Pro
...
-// Clone all nodes with the same ctrl as 'old_ctrl' starting from 'node' by following its inputs. Rewire the cloned nodes
-// to 'new_ctrl'. Returns the clone of 'node'.
-Node* PhaseIdealLoop::clone_nodes_with_same_ctrl(Node* node, ProjNode* old_ctrl, Node* new_ctrl) {
+// Clone all data nodes with a ctrl to the old uncommon projection from `start_node' by following its inputs. Rewire the
+// cloned nodes to the new uncommon projection. Returns the clone of the `start_node`.
+Node* PhaseIdealLoop::clone_nodes_with_same_ctrl(Node* start_node, ProjNode* old_uncommon_proj, Node* new_uncommon_proj) {
+  ResourceMark rm;
   DEBUG_ONLY(uint last_idx = C->unique();)
-  Unique_Node_List nodes_with_same_ctrl = find_nodes_with_same_ctrl(node, old_ctrl);
-  Dict old_new_mapping = clone_nodes(nodes_with_same_ctrl); // Cloned but not rewired, yet
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopPredicate.cpp:250]: Replaced `Dict` with `ResizeableResourceHashtable` which I think is preferable to use.

File: src/hotspot/share/opto/loopnode.hpp [hunk 2]
```diff
@@ -1882,4 +1883,41 @@ class PathFrequency {
...
+  PhaseIdealLoop* const _phase;
+  const Unique_Node_List& _data_nodes;
+  ResizeableResourceHashtable<Node*, Node*, AnyObj::RESOURCE_AREA, mtCompiler> _orig_to_new;
+
+ public:
+  DataNodeGraph(const Unique_Node_List& data_nodes, PhaseIdealLoop* phase)
+      : _phase(phase),
+        _data_nodes(data_nodes),
+        // Use 107 as best guess which is the first resize value in ResizeableResourceHashtable::large_table_sizes.
+        _orig_to_new(107, MaxNodeLimit)
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.hpp:1899]: I'm not sure if this is the right default value - was just a best guess. We usually only have a small number of data nodes to copy.

File: src/hotspot/share/opto/loopnode.hpp [hunk 3]
```diff
@@ -1343,13 +1344,13 @@ class PhaseIdealLoop : public PhaseTransform {
...
-  Node* clone_nodes_with_same_ctrl(Node* node, ProjNode* old_ctrl, Node* new_ctrl);
-  Dict clone_nodes(const Node_List& list_to_clone);
-  void rewire_cloned_nodes_to_ctrl(const ProjNode* old_ctrl, Node* new_ctrl, const Node_List& nodes_with_same_ctrl,
-                                   const Dict& old_new_mapping);
-  void rewire_inputs_of_clones_to_clones(Node* new_ctrl, Node* clone, const Dict& old_new_mapping, const Node* next);
+  const Unique_Node_List& find_nodes_with_same_ctrl(DataInputGraph& data_input_graph, const ProjNode* ctrl);
+  Node* clone_nodes_with_same_ctrl(Node* start_node, ProjNode* old_uncommon_proj, Node* new_uncommon_proj);
+  void fix_cloned_data_node_controls(
+      const ProjNode* old_uncommon_proj, Node* new_uncommon_proj,
+      const ResizeableResourceHashtable<Node*, Node*, AnyObj::RESOURCE_AREA, mtCompiler>& orig_to_new);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.hpp:1353]: Mostly some renaming and adding `const`.

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 4]
```diff
@@ -215,14 +215,16 @@ IfProjNode* PhaseIdealLoop::create_new_if_for_predicate(ParsePredicateSuccessPro
...
 // current ctrl.
-void PhaseIdealLoop::set_ctrl_of_nodes_with_same_ctrl(Node* node, ProjNode* old_ctrl, Node* new_ctrl) {
-  Unique_Node_List nodes_with_same_ctrl = find_nodes_with_same_ctrl(node, old_ctrl);
-  for (uint j = 0; j < nodes_with_same_ctrl.size(); j++) {
-    Node* next = nodes_with_same_ctrl[j];
-    if (next->in(0) == old_ctrl) {
-      _igvn.replace_input_of(next, 0, new_ctrl);
+void PhaseIdealLoop::set_ctrl_of_nodes_with_same_ctrl(Node* start_node, ProjNode* old_uncommon_proj,
+                                                      Node* new_uncommon_proj) {
+  ResourceMark rm;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopPredicate.cpp:220]: Added `ResourceMark`s which I think is now safe after [JDK-8325672](https://bugs.openjdk.org/browse/JDK-8325672).

File: src/hotspot/share/opto/loopnode.hpp [hunk 5]
```diff
@@ -1882,4 +1883,41 @@ class PathFrequency {
...
+  void rewire_clones_to_cloned_inputs();
+
+ public:
+  // Clone the provided data node collection and rewire the clones in such a way to create an identical graph copy.
+  // Set `new_ctrl` as ctrl for the cloned nodes.
+  const ResizeableResourceHashtable<Node*, Node*, AnyObj::RESOURCE_AREA, mtCompiler>& clone(Node* new_ctrl) {
+    clone_nodes(new_ctrl);
+    rewire_clones_to_cloned_inputs();
+    return _orig_to_new;
+  }
```
Comment (eme64) [inline src/hotspot/share/opto/loopnode.hpp:1921]: Currently, it looks like one could call `clone` multiple times. But I think that would be wrong, right?
That is why I'd put all the active logic in the constructor, and only the passive stuff is publicly accessible, with `const` to indicate that these don't have any effect.

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 6]
```diff
@@ -242,61 +244,31 @@ Unique_Node_List PhaseIdealLoop::find_nodes_with_same_ctrl(Node* node, const Pro
...
-    rewire_inputs_of_clones_to_clones(new_ctrl, clone, old_new_mapping, next);
-  }
-}
-
-// Rewire the inputs of the cloned nodes to the old nodes to the new clones.
-void PhaseIdealLoop::rewire_inputs_of_clones_to_clones(Node* new_ctrl, Node* clone, const Dict& old_new_mapping,
-                                                       const Node* next) {
-  for (uint i = 1; i < next->req(); i++) {
-    Node* in = next->in(i);
-    if (!in->is_Phi()) {
```
Comment (eme64) [inline src/hotspot/share/opto/loopPredicate.cpp:291]: What happened with the `is_Phi`? Is it not needed anymore?

File: src/hotspot/share/opto/loopopts.cpp [hunk 7]
```diff
@@ -4495,3 +4495,32 @@ void PhaseIdealLoop::move_unordered_reduction_out_of_loop(IdealLoopTree* loop) {
...
+  _orig_to_new.put(node, clone);
+  _phase->set_ctrl(clone, new_ctrl);
+}
+
+// Rewire the data inputs of all (unprocessed) cloned nodes, whose inputs are still pointing to the same inputs as their
+// corresponding orig nodes, to the newly cloned inputs to create a separate cloned graph.
+void DataNodeGraph::rewire_clones_to_cloned_inputs() {
+  _orig_to_new.iterate_all([&](Node* node, Node* clone) {
+    for (uint i = 1; i < node->req(); i++) {
+      Node** cloned_input = _orig_to_new.get(node->in(i));
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:4519]: You don't need to check for `is_Phi` on `node->in(i)` anymore?

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 8]
```diff
@@ -242,61 +244,31 @@ Unique_Node_List PhaseIdealLoop::find_nodes_with_same_ctrl(Node* node, const Pro
...
   DEBUG_ONLY(uint last_idx = C->unique();)
-  Unique_Node_List nodes_with_same_ctrl = find_nodes_with_same_ctrl(node, old_ctrl);
-  Dict old_new_mapping = clone_nodes(nodes_with_same_ctrl); // Cloned but not rewired, yet
-  rewire_cloned_nodes_to_ctrl(old_ctrl, new_ctrl, nodes_with_same_ctrl, old_new_mapping);
-  Node* clone_phi_input = static_cast<Node*>(old_new_mapping[node]);
-  assert(clone_phi_input != nullptr && clone_phi_input->_idx >= last_idx, "must exist and be a proper clone");
-  return clone_phi_input;
+  const Unique_Node_List nodes_with_same_ctrl = find_nodes_with_same_ctrl(start_node, old_uncommon_proj);
+  DataNodeGraph data_node_graph(nodes_with_same_ctrl, this);
+  auto& orig_to_new = data_node_graph.clone(new_uncommon_proj);
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/loopPredicate.cpp:254]: The beauty of packing it into the constructor is that you have fewer lines here. And that is probably beneficial if you are going to use the class elsewhere -> less code duplication.

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 9]
```diff
@@ -242,61 +244,31 @@ Unique_Node_List PhaseIdealLoop::find_nodes_with_same_ctrl(Node* node, const Pro
...
+  fix_cloned_data_node_controls(old_uncommon_proj, new_uncommon_proj, orig_to_new);
```
Comment (eme64) [inline src/hotspot/share/opto/loopPredicate.cpp:255]: And is there a reason why `fix_cloned_data_node_controls` is not part of the `DataNodeGraph` class? Is there any use of the class where we don't have to call `fix_cloned_data_node_controls`?

**Review by eme64 — CHANGES_REQUESTED** [review]
Looks like a nice refactoring! I left a few comments and questions :)

File: src/hotspot/share/opto/loopnode.hpp (hunk 5 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/loopnode.hpp:1921]: Yes, that would be unexpected, so I agree with you here. But as mentioned earlier, we need to add another method to this class later which does the cloning slightly differently, so we cannot do all the work in the constructor.

We probably have multiple options here:
- Do nothing (could be reasonable as this class is only used rarely and if it's used it's most likely uncommon to clone twice in a row on the same object - and if one does, one probably has a look at the class anyway to notice what to do).
- Add asserts to ensure `clone()` is only called once (adds more code but could be a low overhead option - however, we should think about whether we really want to save the user from itself).
- Return a copy of the hash table and clear it afterward (seems too much overhead for having no such use-case).

I think option 1 and 2 are both fine.

**Review by chhagedorn — COMMENTED** [review]
Thanks for the careful review! I could have added some more comments explaining, what the follow-up refactoring will do to help to better understand some of the decisions in this patch. I've added replies and will update the PR shortly with the mentionied changes.

**Comment by eme64** [comment]
@chhagedorn ah ok, I see. I didn't quite realize how you were going to extend the code later before your comments. In that case you can of course leave the computations outside the constructor. We can still discuss the final shape of the code once you do the next RFE's on the same code :)

I'll wait for your code updates to re-review, just ping me ;)

File: src/hotspot/share/opto/loopnode.hpp [hunk 10]
```diff
@@ -35,6 +35,7 @@
 class CmpNode;
 class BaseCountedLoopEndNode;
 class CountedLoopNode;
+class DataInputGraph;
```
Comment (eme64) [inline src/hotspot/share/opto/loopnode.hpp:38]:
```suggestion
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 11]
```diff
@@ -4495,3 +4495,32 @@ void PhaseIdealLoop::move_unordered_reduction_out_of_loop(IdealLoopTree* loop) {
     assert(phi->outcnt() == 1, "accumulator is the only use of phi");
   }
 }
+
+// Clone all nodes in _data_nodes.
+void DataNodeGraph::clone_nodes(Node* new_ctrl) {
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:4500]:
```suggestion
void DataNodeGraph::clone_data_nodes(Node* new_ctrl) {
```
Then the comment would be obsolete

File: src/hotspot/share/opto/loopopts.cpp (hunk 7 above)
Comment (eme64) [reply src/hotspot/share/opto/loopopts.cpp:4519]: Got it, great. I just looked for a matching `is_Phi` in your diff and did not find it. But it is already covered in existing code, great!

File: src/hotspot/share/opto/loopnode.hpp (hunk 10 above)
Comment (eme64) [reply src/hotspot/share/opto/loopnode.hpp:38]: Is also dead

File: src/hotspot/share/opto/replacednodes.cpp [hunk 12]
```diff
@@ -208,7 +208,7 @@ void ReplacedNodes::apply(Compile* C, Node* ctl) {
       hash_table_size++;
     }
     // Map from current node to cloned/replaced node
-    ResizeableResourceHashtable<Node*, Node*, AnyObj::RESOURCE_AREA, mtCompiler> clones(hash_table_size, hash_table_size);
+    OrigToNewHashtable clones(hash_table_size, hash_table_size);
```
Comment (eme64) [inline src/hotspot/share/opto/replacednodes.cpp:211]: Nice.
Not your problem here. But should there not be a ResouceMark before this hashtable? There is one at the beginning of the function, but we create many of these hashtables in a loop, without any ResourceMarks in between reclaiming the memory...
A but then the hashmaps and stack/to_fix etc would allocate from the ResourceArea, but start at different ResourceMarks... bad idea. Hmm.

File: src/hotspot/share/opto/loopnode.hpp (hunk 5 above)
Comment (eme64) [reply src/hotspot/share/opto/loopnode.hpp:1919]: Could you add asserts that the `_orig_to_new` is empty before we clone? That would be a check that nothing was cloned yet, and we do not accidentally mix up two clone operations.

**Review by eme64 — CHANGES_REQUESTED** [review]
Nice, looks better already :)

**Review by eme64 — APPROVED** [review]
Nice refactoring, looking forward to your next PR's on this ;)

**Comment by chhagedorn** [comment]
Thanks Emanuel for your review and comments! I'll send it out after this one goes in  :-)

File: src/hotspot/share/opto/replacednodes.cpp (hunk 12 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/replacednodes.cpp:211]: As discussed offline, we should probably go over all uses of resource allocated things like `Node_Lists`, `Node_Stack` etc. at some point and check if there are missing resource marks.

**Review by TobiHartmann — APPROVED** [review]
That looks good to me.

**Comment by chhagedorn** [comment]
Thanks for your review Tobias!

/integrate

**Pull request closed without merging**